inc/migrate-images.php

```diff
@@ -0,0 +1,177 @@
+	 *		'url' string
+	 * 		'year' string
+	 *		'month' string
+	 *		'filename' string
+	 * 		'result' bool
+	 */
+	protected $images = array();
+
+	/**
+	 * @var array Stats of results
+	 */
+	protected $results = array( 'success' => 0, 'error' => 0, 'skip' => 0 );
+
+	/**
+	 * @var int|null Post ID being migrated
+	 */
+	protected $id;
+
+	/**
+	 * @var string Regex pattern to match migrateable images
+	 */
+	protected $pattern = '/https?:\/\/[\w\.]*hackshackers\.com\/wp-content\/uploads\/(\d+)\/(\d+)\/([^)\s\n"]+)/';
+
+	/**
+	 * @var string Destination directory for images to migrate
+	 */
+	protected $dest_dir = '';
+
+	/**
+	 * @var string Source directory for images to migrate
+	 */
+	protected $source_dir = '';
+
+	/**
+	 * Instantiate and migrate
+	 *
+	 * @param string $markdown Markdown content to parse for migratable images
+	 * @param bool $dry_run If true, mock migration only. Defaults to false.
+	 */
+	function __construct( $markdown, $dry_run = false ) {
+		$content_dir = defined( 'HH_HUGO_UNIT_TESTS_RUNNING' ) ? 'hugo-images-test' : 'hugo-images';
+		$this->dest_dir = trailingslashit( HH_HUGO_COMMAND_DIR ) . $content_dir;
+		$this->source_dir = wp_upload_dir()['basedir'];
+
+		// just return the class for unit testing
+		if ( empty( $markdown ) && defined( 'HH_HUGO_UNIT_TESTS_RUNNING' ) ) {
+			return $this;
+		}
+
+		$images = $this->extract_images( $markdown );
+		if ( empty( $images ) ) {
+			return;
+		}
+		$this->images = $images;
+
+
+		foreach ( $this->images as &$image ) {
+			if ( $this->should_migrate_image( $image ) ) {
+				$image['result'] = $dry_run ? 'success' : $this->migrate_image( $image, $dry_run );
+				if ( 'success' === $image['result'] ) {
+					$this->results['success']++;
+				} else {
+					$this->results['error']++;
+				}
+			} else {
+				$image['result'] = 'skip';
+				$this->results['skip']++;
+			}
+		}
+	}
+
+	/**
+	 * Getter for migration results
+	 *
+	 * @return array
+	 */
+	public function results() {
+		return array(
+			'results' => $this->results,
+			'images' => $this->images,
+		);
+	}
+
+	/**
+	 * Extra images data from Markdown
+	 *
+	 * @param string $input Markdown input
+	 * @return array Images data
+	 */
+	public function extract_images( $markdown ) {
+		preg_match_all( $this->pattern, $markdown, $matches, PREG_SET_ORDER );
+		return array_map( function( $match ) {
+			return array(
+				'url' => $match[0],
+				'year' => $match[1],
+				'month' => $match[2],
+				'filename' => $match[3],
+			);
+		}, $matches );
+	}
+
+	/**
+	 * Get absolute path for image
+	 *
+	 * @param array $image Image data
+	 * @param string $type Path type, defaults to 'rel', can also use 'src' or 'dest'
+	 * @return string Path
+	 */
+	public function image_path( $image, $type = 'rel' ) {
+		$rel_path = sprintf(
+			'%s/%s/%s',
+			$image['year'],
+			$image['month'],
+			$image['filename']
+		);
+
+		switch ( $type ) {
+			case 'rel':
+			default:
```

Comment:
It technically doesn't have to be. In this case I put it up there with `'rel'` because I wanted to put the expected default value for `$type` first, and it made sense (to me at least) to put the switch statement's default next to the variable's default.